config/config.go

```diff
@@ -82,6 +91,98 @@ func (c *Client) prefixed(strs ...string) string {
 	return str
 }
 
+func writeNode(node *client.Node, writer *tar.Writer, baseDirectoryName string) error {
+	now := time.Now()
+
+	header := &tar.Header{
+		AccessTime: now,
+		ChangeTime: now,
+		ModTime:    now,
+		Name:       baseDirectoryName + node.Key,
+	}
+
+	if node.Dir {
+		header.Mode = 0700
+		header.Typeflag = tar.TypeDir
+	} else {
+		header.Mode = 0600
+		header.Size = int64(len(node.Value))
+		header.Typeflag = tar.TypeReg
+	}
+
+	err := writer.WriteHeader(header)
+	if err != nil {
+		return errored.Errorf("Failed to write tar entry header: %v", err)
+	}
+
+	// we don't have to write anything for directories except the header
+	if !node.Dir {
+		_, err = writer.Write([]byte(node.Value))
+		if err != nil {
+			return errored.Errorf("Failed to write tar entry: %v", err)
+		}
+	}
+
+	for _, n := range node.Nodes {
+		writeNode(n, writer, baseDirectoryName)
+	}
+
+	return nil
+}
+
+// DumpTarball dumps all the keys under the current etcd prefix into a
+// gzip'd tarball'd directory-based representation of the namespace.
+func (c *Client) DumpTarball() (string, error) {
+	resp, err := c.etcdClient.Get(context.Background(), c.prefix, &client.GetOptions{Sort: true, Recursive: true, Quorum: true})
+
+	if err != nil {
+		return "", errored.Errorf(`Failed to recursively GET "%s" namespace from etcd: %v`, c.prefix, err)
+	}
+
+	now := time.Now()
+
+	// tar hangs during unpacking if the base directory has colons in it
+	// unless --force-local is specified, so use the simpler "%Y%m%d-%H%M%S".
+	niceTimeFormat := fmt.Sprintf("%d%02d%02d-%02d%02d%02d",
+		now.Year(), now.Month(), now.Day(),
+		now.Hour(), now.Minute(), now.Second())
+
+	file, err := ioutil.TempFile("", "etcd_dump_"+niceTimeFormat+"_")
+	if err != nil {
+		return "", errored.Errorf("Failed to create tempfile: %v", err)
+	}
+	defer file.Close()
+
+	// create a gzipped, tarballed, buffered writer which self-closes
+	gzipWriter := gzip.NewWriter(file)
+	defer gzipWriter.Close()
+
+	bufWriter := bufio.NewWriter(gzipWriter)
+	defer bufWriter.Flush()
+
+	tarWriter := tar.NewWriter(bufWriter)
+	defer tarWriter.Close()
```

Comment:
Hmm. How do you feel about a single `defer` that encapsulates all the close/flush functionality?

I think this will be cleaner because there will be one defer and it will be less confusing which one fires first.

Alternatively, please comment it if you object.

I have no strong opinion, I just wanted to offer an idea.